azure_monitor/src/azure_monitor/sdk/auto_collection/dependency_metrics.py

```diff
@@ -1,16 +1,53 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
+import logging
+import threading
 import time
 from typing import Dict
 
+import requests
+from opentelemetry import context
 from opentelemetry.metrics import Meter, Observer
 from opentelemetry.sdk.metrics import UpDownSumObserver
 
-from azure_monitor.sdk.auto_collection.metrics_span_processor import (
-    AzureMetricsSpanProcessor,
-)
-
+_dependency_lock = threading.Lock()
+logger = logging.getLogger(__name__)
 dependency_map = dict()
+ORIGINAL_REQUEST = requests.Session.request
+
+
+def dependency_patch(*args, **kwargs) -> None:
+    start_time = time.time()
+
+    try:
+        result = ORIGINAL_REQUEST(*args, **kwargs)
+    except Exception as exc:  # pylint: disable=broad-except
+        exception = exc
+        result = getattr(exc, "response", None)
+    end_time = time.time()
+
+    # Only collect request metric if sent from non-exporter thread
+    if context.get_value("suppress_instrumentation") is None:
+        # We don't want multiple threads updating this at once
+        with _dependency_lock:
```

Comment:
This will actually block the main thread until the dependency count processing is finished. Is there some way we can return the result but make the count logic asynchronous?

Reply:
Discussed this offline, we will follow up on improving locks in a different PR

Reply:
Can you create an issue to track this?